samples/nrf9160/modem_shell/src/location/location_srv_ext_nrf_cloud.c

```diff
@@ -0,0 +1,259 @@
+				      pgps_req->gps_time_of_day);
+	if (ret == NULL) {
+		mosh_error("Failed to add GPS time to P-GPS request");
+		err = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Convert CJSON object to string and send to nRF Cloud */
+	msg_string = cJSON_PrintUnformatted(pgps_req_obj);
+	if (msg_string == NULL) {
+		mosh_error("Could not allocate memory for request message");
+		goto cleanup;
+	}
+
+	struct nrf_cloud_tx_data mqtt_msg = {
+		.data.ptr = msg_string,
+		.data.len = strlen(msg_string),
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+	};
+
+	err = nrf_cloud_send(&mqtt_msg);
+	if (!err) {
+		mosh_print("P-GPS data requested");
+	}
+
+cleanup:
+	cJSON_Delete(pgps_req_obj);
+	cJSON_free(msg_string);
+
+	if (err) {
+		mosh_error("nRF Cloud P-GPS request failed, error: %d", err);
+	}
+}
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR) || defined(CONFIG_LOCATION_METHOD_WIFI)
+static void location_srv_ext_nrf_cloud_location_ready_cb(
+	const struct nrf_cloud_location_result *const result)
+{
+	if ((result != NULL) && (result->err == NRF_CLOUD_ERROR_NONE)) {
+		nrf_cloud_location.latitude = result->lat;
+		nrf_cloud_location.longitude = result->lon;
+		nrf_cloud_location.accuracy = (double)result->unc;
+
+		k_sem_give(&location_ready);
+	} else {
+		if (result) {
+			mosh_error(
+				"Unable to determine location from cloud response, error: %d",
+				result->err);
+		}
+		/* Reset the semaphore to unblock location_srv_ext_nrf_cloud_pos_get()
+		 * and make it return an error.
+		 */
+		k_sem_reset(&location_ready);
+	}
+}
+
+static int location_srv_ext_nrf_cloud_pos_get(
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	const struct lte_lc_cells_info *cell_data,
+#else
+	void *cell_data,
+#endif
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	const struct wifi_scan_info *scanning_results,
+#else
+	void *scanning_results,
+#endif
+	struct location_data *const location)
+{
+	int err;
+	nrf_cloud_location_response_t callback = NULL;
+
+	k_sem_reset(&location_ready);
+
+	if (location) {
+		callback = location_srv_ext_nrf_cloud_location_ready_cb;
+	}
+
+	mosh_print("Sending positioning request to cloud via MQTT");
+	err = nrf_cloud_location_request(
+		cell_data, scanning_results, (callback != NULL), callback);
+	if (err == -EACCES) {
+		mosh_error("Cloud connection is not established");
+		return err;
+	} else if (err) {
+		mosh_error("Failed to request positioning data, error: %d", err);
+		return err;
+	}
+
+	if (callback) {
+		/* We are in the middle of the Cellular/Wi-Fi positioning and we've received
+		 * an event from Location library for handling the cloud transaction.
+		 * Location library has stopped it's method specific timeout.
+		 * We could set cellular_timeout or wifi_timeout into the timeout of k_sem_take()
+		 * but we have opted to simplify implementation of the application and use
+		 * a separate timeout for a cloud transaction.
```

Comment:
```suggestion
		/* We are in the middle of the Cellular/Wi-Fi positioning and we have received
		 * an event from Location library for handling the cloud transaction.
		 * Location library has stopped its method-specific timeout.
		 * We could set cellular_timeout or wifi_timeout into the timeout of k_sem_take()
		 * but we have opted to simplify implementation of the application and use
		 * a separate timeout for a cloud transaction.
```
In general, you should not talk about we. But this is code, so totally up to you.